### ISSUE TITLE
Flatpak and Environment variables

### ISSUE THREAD

**Issue by CrashyBang** [flatpak/flatpak#2464]
## Linux distribution and version
```
Void Linux
```
## Flatpak version
```
Flatpak 1.0.6
```
## Description of the problem
So to get all my flatpak applications (Discord/Spotify/Slack) working I am having to export a range of variables, as follows:
1) DBUS:
```
# .zshrc
export $(dbus-launch)
```
Without this I get errors a long the lines of:
```
(discord:4): libappindicator-WARNING **: 13:25:57.466: Unable to get the session bus: Error spawning command line ?dbus-launch --autolaunch=22f4463094b38acf73066d0d5c17834d --binary-syntax --close-stderr?: Child process exited with code 1
```
2) Pulseaudio
```
# .zshrc
export PULSE_SERVER=unix:/tmp/pulse-PKdhtXMmr18n/native
```
Without this I get errors a long the lines of:
```
ALSA lib ../../pulse/pulse.c:243:(pulse_connect) PulseAudio: Unable to connect: Connection refused

ALSA lib ../../src/dlmisc.c:170:(snd_dlsym_verify) unable to verify version for symbol _snd_pcm_plug_open
ALSA lib ../../src/dlmisc.c:296:(snd1_dlobj_cache_get) symbol _snd_pcm_plug_open is not defined inside [builtin]
```
3) XDG
```
# /etc/profile
export XDG_RUNTIME_DIR=/run/user/$(id -u)
```
Without this I get errors a long the lines of:
```
sh: xdg-mime: command not found
```

What I want to know is what each of these variables mean (I scoured them from various parts of the web notably issue [146](https://github.com/moonlight-stream/moonlight-qt/issues/146) and this [comment](https://github.com/flatpak/flatpak/issues/2169#issuecomment-448652037) and if they are necessary? (What I mean by this is should they be fixed on the flatpak side or should they be fixed on my side like they are now).

Side note `export $(dbus-launch)` makes the flatpak applications take a lot longer to load but no errors.

**UPDATE**
```
sh: xdg-mime: command not found
```
Is back and I have no idea why, I was fiddling around with shells trying to see if I could get the variables exported everywhere in an easy way and it started showing up again even with:
```
#/etc/profile
export XDG_RUNTIME_DIR=/run/user/$(id -u)
```
So I am starting to think these two are unrelated and it may not even be necessary to have.

**Comment by kirbyfan64** [comment]
> DBUS

D-Bus is a message bus, and it's generally required for a lot of stuff in the Linux world. TBH it's a miracle that a lot of stuff even functioned for you without it... This is a Discord/Spotify/Slack requirement (due to the use of libappindicator by Electron) and shouldn't have anything to do with Flatpak. 

> Pulseaudio

What's the normal value of `$PULSE_SERVER` for you? That variable is used to locate the Pulseaudio server communication socket. 

FWIW Pulse itself is required by Discord, I believe due to (again) Electron?

> XDG
`$XDG_RUNTIME_DIR` is used as a user-specific temporary directory (primarily holding things like sockets), usually under /run (that way there's no problems with stuff from multiple users conflicting inside /tmp). 

As for the xdg-mime issue, that's entirely harmless. Flatpak (IIRC intentionally) doesn't include xdg-mime, which throws Discord and some other Electron apps off a bit, but it still functions correctly.

**Comment by matthiasclasen** [comment]
> export XDG_RUNTIME_DIR=/run/user/$(id -u)

The question is, why are you doing all these things by hand ? Like all the other things you mention, XDG_RUNTIME_DIR is part of the session setup and you should not have to mess with it manually.

**Comment by CrashyBang** [comment]
Hey @matthiasclasen well if I don't do the DBUS and Pulseaudio stuff by hand the applications simply don't work.

As for the `XDG_RUNTIME_DIR` I don't entirely understand it, but from what I understand it is not being set?

**Comment by kirbyfan64** [comment]
`export $(dbus-launch)` starts D-Bus, then it exports the environment variables needed for it to be used (it prints them, which is then passed to export). 

As for `$PULSE_SERVER`...that's pretty bizarre. PulseAudio generally picks a different directory each time the system is booted, and your system *should* be exporting it. Maybe ask around the Void Linux forums on how PulseAudio is started and why it doesn't expose `$PULSE_SERVER`. 

Also, are you sure you set them up correctly? Did you follow [these directions](https://wiki.voidlinux.eu/PulseAudio#Recommended_setup)?

`xdg-mime` errors are probably going to be entirely random and unpredictable, and they're definitely unfixable (at least, not without some pretty pointless tweaking). 

For XDG_RUNTIME_DIR, it seems to be a Void issue. I found [this](https://github.com/voidlinux/void-packages/issues/9812) similar problem with no solution. What if you add the export command for it to /etc/profile? (After making sure you tried what I mentioned under PulseAudio, of course.)

**Comment by CrashyBang** [comment]
Hey Kirby,

I hadn't followed those directions I had just installed `alsa-utils` and `pulseaudio` and run `pulseaudio -D` but even after trying them still no dice, I will hop on the void IRC and see if I can figure out what is going on here (I will comment here if I find anything out).

I originally had `XDG_RUNTIME_DIR` in `/etc/profile` but from what I can tell it was not having an effect (as in if I removed it it did not seem to change anything) may I ask how it and the `$PULSE_SERVER` relate?

Cheers.
